api/apiclient_test.go

```diff
@@ -561,6 +572,128 @@ func (s *apiclientSuite) TestOpenWithRedirect(c *gc.C) {
 	})
 }
 
+func (s *apiclientSuite) TestOpenCachesDNS(c *gc.C) {
+	fakeDialer := func(ctx context.Context, urlStr string, tlsConfig *tls.Config, ipAddr string) (jsoncodec.JSONConn, error) {
+		return fakeConn{}, nil
+	}
+	dnsCache := make(dnsCacheMap)
+	conn, err := api.Open(&api.Info{
+		Addrs: []string{
+			"place1.example:1234",
+		},
+		SkipLogin: true,
+		CACert:    jtesting.CACert,
+	}, api.DialOpts{
+		Timeout:       5 * time.Second,
+		RetryDelay:    1 * time.Second,
+		DialWebsocket: fakeDialer,
+		IPAddrResolver: fakeResolver{
+			"place1.example": {"0.1.1.1"},
+		},
+		DNSCache: dnsCache,
+		Clock:    &fakeClock{},
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(conn, gc.NotNil)
+	c.Assert(dnsCache.Lookup("place1.example"), jc.DeepEquals, []string{"0.1.1.1"})
+}
+
+func (s *apiclientSuite) TestDNSCacheUsed(c *gc.C) {
+	var dialed string
+	fakeDialer := func(ctx context.Context, urlStr string, tlsConfig *tls.Config, ipAddr string) (jsoncodec.JSONConn, error) {
+		dialed = ipAddr
+		return fakeConn{}, nil
+	}
+	conn, err := api.Open(&api.Info{
+		Addrs: []string{
+			"place1.example:1234",
+		},
+		SkipLogin: true,
+		CACert:    jtesting.CACert,
+	}, api.DialOpts{
+		Timeout:       5 * time.Second,
+		RetryDelay:    1 * time.Second,
+		DialWebsocket: fakeDialer,
+		IPAddrResolver: fakeResolver{
+			"place1.example": {"0.2.2.2"},
+		},
+		DNSCache: dnsCacheMap{
+			"place1.example": {"0.1.1.1"},
+		},
+		Clock: &fakeClock{},
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(conn, gc.NotNil)
+	// The dialed IP address should have come from the cache, not the IP address
+	// resolver.
+	c.Assert(dialed, gc.Equals, "0.1.1.1:1234")
+	c.Assert(conn.IPAddr(), gc.Equals, "0.1.1.1:1234")
+}
+
+func (s *apiclientSuite) TestFallbackToIPLookupWhenCacheOutOfDate(c *gc.C) {
+	var mu sync.Mutex
+	dialed := make(map[string]bool)
+	fakeDialer := func(ctx context.Context, urlStr string, tlsConfig *tls.Config, ipAddr string) (jsoncodec.JSONConn, error) {
+		mu.Lock()
+		defer mu.Unlock()
+		dialed[ipAddr] = true
+		if ipAddr == "0.2.2.2:1234" {
+			return fakeConn{}, nil
+		}
+		return nil, errors.Errorf("bad address")
+	}
+	dnsCache := dnsCacheMap{
+		"place1.example": {"0.1.1.1"},
+	}
+	conn, err := api.Open(&api.Info{
+		Addrs: []string{
+			"place1.example:1234",
+		},
+		SkipLogin: true,
+		CACert:    jtesting.CACert,
+	}, api.DialOpts{
+		Timeout:       5 * time.Second,
```

Comment:
does this mean that this particular test takes 5s to retry 5 times before deciding to fallback to resolver instead of using the local cache?
Can we make it a bit faster (100ms/500ms)?
Maybe it doesn't because we also pass in fakeClock?